pandora-server-directory/src/networking/connection_client.ts

```diff
@@ -63,12 +72,45 @@ export class ClientConnection extends IncomingConnection<IDirectoryClient, IClie
 			this._account = null;
 		}
 		if (account) {
+			Assert(token?.reason === AccountTokenReason.LOGIN);
 			account.touch();
 			this._account = account;
+			this._loginTokenEventUnsubscribe?.();
+			this._loginToken = token;
+			this._loginTokenEventUnsubscribe = token.onAny((ev) => this.onTokenEvent(ev));
 			this.joinRoom(account.associatedConnections);
 		}
 	}
 
+	private onTokenEvent(event: Partial<TypedEventEmitterEvents<AccountToken>>): void {
+		if (event.tokenDestroyed != null) {
+			Assert(this._loginToken === event.tokenDestroyed);
+			this.onAccountTokenDestroyed();
+		}
+		if (event.extended != null) {
+			Assert(this._loginToken === event.extended);
+			this.sendMessage('loginTokenChanged', {
+				value: event.extended.value,
+				expires: event.extended.expires,
+			});
```

Comment:
This is slightly problematic; imagine the following case:
- You have two tabs open
- In one of the tabs you are extending token, the other tab is on the background
- As the other tab is on the background it can easily DC
- The token refreshes, the current tab does receive token-updating event
- The other tab then reconnects, but it won't receive this and still will have the old data.

I'm not fully sure how to handle this. One possibility I see is having a shared store for the token in `localStorage`, as all open tabs share the value and get update when the saved value changes. In that case only the tab updating the token would store the new data in the store and everyone will see it.

Reply:
it is already shared:
```ts
private readonly _authToken = BrowserStorage.create<AuthToken | undefined>('authToken', undefined, AuthTokenSchema);
```